GMS/FullGMS.cpp

```diff
@@ -0,0 +1,193 @@
+    UA_NODEID_NUMERIC(nsFromUri(m_pServer, constants::NsGMSUri), UA_GMSID_GMSTYPE),
+    objAttr,
+    nullptr,
+    m_nodeId.NodeId);
+
+  UA_ObjectAttributes_clear(&objAttr);
+  UmatiServerLib::Bind::MembersRefl(gms, m_pServer, m_nodeId, n);
+
+  InstantiateIdentification();
+  InstantiateMonitoring();
+  InstantiateProduction();
+  InstantiateResultManagement();
+  InstantiateEquipment();
+  InstantiateNotification();
+}
+
+void FullGMS::InstantiateIdentification() {
+  {
+    std::stringstream ss;
+    ss << "http://www.isw.uni-stuttgart.de/#" << MachineName;
+    gms.Identification->ProductInstanceUri = ss.str();
+  }
+
+  InstantiateOptional(gms.Identification->YearOfConstruction, m_pServer, n);
+  InstantiateOptional(gms.Identification->ProductCode, m_pServer, n);
+  InstantiateOptional(gms.Identification->SoftwareRevision, m_pServer, n);
+  InstantiateOptional(gms.Identification->DeviceClass, m_pServer, n);
+  InstantiateOptional(gms.Identification->Location, m_pServer, n);
+  InstantiateOptional(gms.Identification->Model, m_pServer, n);
+
+  gms.Identification->Manufacturer = {"", "ISW University of Stuttgart"};
+  gms.Identification->ProductCode = "CMM_1234";
+  gms.Identification->YearOfConstruction = 2022;
+  gms.Identification->SoftwareRevision = "v1.05";
+  gms.Identification->DeviceClass = "CoordinateMeasuringMachine";
+  gms.Identification->Location = "CTRL 5 5318/VIRTUAL 0 0/N 48.7685303 E 9.2653923";
+  gms.Identification->Model = {"", MachineName};
+
+  {
+    std::hash<std::string> hasher;
+    std::stringstream ss;
+    ss << "2022-" << hasher(MachineName);
+    gms.Identification->SerialNumber = ss.str();
+  }
+
+  InstantiateOptional(gms.Identification->SubDeviceClass, m_pServer, n);
+  InstantiateOptional(gms.Identification->Workspace, m_pServer, n);
+
+  gms.Identification->SubDeviceClass = "CoordinateMeasuringMachine";
+}
+
+void FullGMS::InstantiateMonitoring() {
+  InstantiateOptional(gms.Monitoring->MachineTool->PowerOnDuration, m_pServer, n);
+  InstantiateOptional(gms.Monitoring->MachineTool->FeedOverride, m_pServer, n);
+  gms.Monitoring->MachineTool->OperationMode = UA_MachineOperationMode::UA_MACHINEOPERATIONMODE_AUTOMATIC;
+}
+
+void FullGMS::InstantiateProduction() {
+  InstantiateOptional(gms.Production->ActiveProgram->State, m_pServer, n);
+  gms.Production->ActiveProgram->NumberInList = 0;
+  gms.Production->ActiveProgram->Name = "Basic Measurment Program";
+  gms.Production->ActiveProgram->State->CurrentState->Value = {"en", "Running"};
+  gms.Production->ActiveProgram->State->CurrentState->Number = 1;
+  gms.Production->ActiveProgram->State->CurrentState->Id =
+    UA_NODEID_NUMERIC(nsFromUri(m_pServer, constants::NsMachineToolUri), UA_MACHINETOOLID_PRODUCTIONSTATEMACHINETYPE_RUNNING);
+}
+
+void FullGMS::InstantiateEquipment() {
+  InstantiateOptional(gms.Equipment->Tools, m_pServer, n);
+
+  auto &tool1 = gms.Equipment->Tools->Tool.Add<machineTool::Tool_t>(m_pServer, n, {m_nsIndex, "Tool 1"});
```

Comment:
Why do you add an machine tool toll and a gms sensor?

Reply:
Ah, the idea was to show that the Tool list can contain both MT tools and GMS sensors but ofc GMSSensor should normally be used since it is a subtype of MT Tool. I can remove this if you think it's unnecessary.

Reply:
Oh I just realized that the GMS EquipmentType extends the MT EquipmentType by adding an optional "Sensors" folder for all the sensors. Should the sensors be put in there instead of under the "Tools" folder? Is the "Sensors" folder supposed to replace the "Tools" folder? Or is the "Sensors" folder the container for the "AdditionalSensor" types?

The specification document is showing some inconsistencies that are confusing me 😅

Reply:
I removed the MT Tool instance